{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "33130f65-0c5e-44eb-b00c-c56fa3f8e792",
   "metadata": {},
   "source": [
    "# Get resources from bio.tools\n",
    "\n",
    "See the [bio.tools API reference](https://biotools.readthedocs.io/en/latest/api_reference.html) and [API usage guide](https://biotools.readthedocs.io/en/latest/api_usage_guide.html) on how to query [bio.tools](https://bio.tools/) records.\n",
    "\n",
    "JSON schema for bio.tools: https://github.com/bio-tools/biotoolsSchema/blob/master/jsonschema/biotoolsj.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1c64a56-5656-4b8e-a290-4a756727e9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "from pathlib import Path\n",
    "from motbxtools import motbxschema\n",
    "import pprint\n",
    "pp = pprint.PrettyPrinter(indent = 2)\n",
    "\n",
    "CWD = Path.cwd()\n",
    "if CWD.name != \"notebooks\":\n",
    "    print(\"Make sure to run this notebook from the 'notebooks' directory.\")\n",
    "\n",
    "MOTBX_DIR = CWD.parent # home directory of this repository\n",
    "# path to directory where resources YAML file are saved\n",
    "RESOURCES_DIR = MOTBX_DIR.joinpath(\"resources/external/biotools\")\n",
    "RESOURCES_DIR.mkdir(parents=True, exist_ok=True)\n",
    "# path to JSON SCHEMA file defining structure of MOTBX resources\n",
    "SCHEMA_JSON = MOTBX_DIR.joinpath(\"schema/motbxschema.json\")\n",
    "# load MOTBX schema used for validation\n",
    "schema = motbxschema.MotbxSchema(SCHEMA_JSON)\n",
    "\n",
    "# bio.tools REST API\n",
    "base_url = \"https://bio.tools/api\"\n",
    "sign_in_url = f\"{base_url}/rest-auth/login/\" # for submitting resources\n",
    "\n",
    "headers = {\n",
    "  'Accept': 'application/json',\n",
    "  'Content-Type': 'application/json'\n",
    "}\n",
    "\n",
    "url = f\"{base_url}/tool/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "510b58f6",
   "metadata": {},
   "source": [
    "## Perform some test queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "18d90391-204d-4620-bbd6-09516ff09b39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['count', 'next', 'previous', 'list'])\n",
      "1\n",
      "{ 'accessibility': 'Open access',\n",
      "  'additionDate': '2015-12-02T11:00:49Z',\n",
      "  'biotoolsCURIE': 'biotools:molgenis',\n",
      "  'biotoolsID': 'molgenis',\n",
      "  'collectionID': [ 'NBIC',\n",
      "                    'Rare Disease',\n",
      "                    'BBMRI',\n",
      "                    'BioMedBridges Tools',\n",
      "                    'ELIXIR-NL'],\n",
      "  'community': None,\n",
      "  'confidence_flag': None,\n",
      "  'cost': 'Free of charge',\n",
      "  'credit': [ { 'email': None,\n",
      "                'fundrefid': None,\n",
      "                'gridid': None,\n",
      "                'name': 'UMC Groningen and collaborators',\n",
      "                'note': None,\n",
      "                'orcidid': None,\n",
      "                'rorid': None,\n",
      "                'typeEntity': 'Institute',\n",
      "                'typeRole': ['Provider'],\n",
      "                'url': 'http://www.rug.nl/'},\n",
      "              { 'email': None,\n",
      "                'fundrefid': None,\n",
      "                'gridid': None,\n",
      "                'name': 'BioMedBridges',\n",
      "                'note': None,\n",
      "                'orcidid': None,\n",
      "                'rorid': None,\n",
      "                'typeEntity': 'Consortium',\n",
      "                'typeRole': ['Documentor'],\n",
      "                'url': 'http://www.biomedbridges.eu'},\n",
      "              { 'email': None,\n",
      "                'fundrefid': None,\n",
      "                'gridid': None,\n",
      "                'name': 'ELIXIR-NL',\n",
      "                'note': None,\n",
      "                'orcidid': None,\n",
      "                'rorid': None,\n",
      "                'typeEntity': None,\n",
      "                'typeRole': [],\n",
      "                'url': 'https://www.dtls.nl/elixir-nl/elixir-nl-2/'},\n",
      "              { 'email': 'molgenis-hackers@googlegroups.com',\n",
      "                'fundrefid': None,\n",
      "                'gridid': None,\n",
      "                'name': None,\n",
      "                'note': None,\n",
      "                'orcidid': None,\n",
      "                'rorid': None,\n",
      "                'typeEntity': 'Person',\n",
      "                'typeRole': ['Primary contact'],\n",
      "                'url': None},\n",
      "              { 'email': 'm.a.swertz@rug.nl',\n",
      "                'fundrefid': None,\n",
      "                'gridid': None,\n",
      "                'name': 'Morris Swertz',\n",
      "                'note': None,\n",
      "                'orcidid': None,\n",
      "                'rorid': None,\n",
      "                'typeEntity': 'Person',\n",
      "                'typeRole': ['Primary contact'],\n",
      "                'url': None},\n",
      "              { 'email': 'molgenis-support@umcg.nl',\n",
      "                'fundrefid': None,\n",
      "                'gridid': None,\n",
      "                'name': None,\n",
      "                'note': None,\n",
      "                'orcidid': None,\n",
      "                'rorid': None,\n",
      "                'typeEntity': 'Person',\n",
      "                'typeRole': ['Support'],\n",
      "                'url': None}],\n",
      "  'description': 'A framework for scientific data management, exploration, '\n",
      "                 'integration and analysis. It is adaptable to many types of '\n",
      "                 'scientific data and data workflows, and there are several '\n",
      "                 'pre-configured MOLGENIS systems for specific data types.',\n",
      "  'documentation': [ { 'note': None,\n",
      "                       'type': ['General'],\n",
      "                       'url': 'https://molgenis.gitbooks.io/molgenis/content/'}],\n",
      "  'download': [ { 'note': None,\n",
      "                  'type': 'Source code',\n",
      "                  'url': 'https://github.com/molgenis/molgenis',\n",
      "                  'version': None}],\n",
      "  'editPermission': {'authors': ['ELIXIR-EE'], 'type': 'group'},\n",
      "  'elixirCommunity': [],\n",
      "  'elixirNode': ['Netherlands'],\n",
      "  'elixirPlatform': ['Interoperability'],\n",
      "  'elixir_badge': 0,\n",
      "  'function': [ { 'cmd': None,\n",
      "                  'input': [],\n",
      "                  'note': None,\n",
      "                  'operation': [ { 'term': 'Data retrieval',\n",
      "                                   'uri': 'http://edamontology.org/operation_2422'},\n",
      "                                 { 'term': 'Aggregation',\n",
      "                                   'uri': 'http://edamontology.org/operation_3436'}],\n",
      "                  'output': []}],\n",
      "  'homepage': 'https://molgenis.org',\n",
      "  'homepage_status': 0,\n",
      "  'language': ['R', 'Java', 'SQL'],\n",
      "  'lastUpdate': '2020-06-16T10:55:23Z',\n",
      "  'license': 'LGPL-3.0',\n",
      "  'link': [],\n",
      "  'maturity': 'Mature',\n",
      "  'name': 'MOLGENIS',\n",
      "  'operatingSystem': ['Linux', 'Windows', 'Mac'],\n",
      "  'otherID': [],\n",
      "  'owner': 'mswertz',\n",
      "  'publication': [ { 'doi': '10.1093/bioinformatics/btw155',\n",
      "                     'metadata': { 'abstract': '© 2016 The Author 2016. '\n",
      "                                               'Published by Oxford University '\n",
      "                                               'Press.Motivation: While the '\n",
      "                                               'size and number of biobanks, '\n",
      "                                               'patient registries and other '\n",
      "                                               'data collections are '\n",
      "                                               'increasing, biomedical '\n",
      "                                               'researchers still often need '\n",
      "                                               'to pool data for statistical '\n",
      "                                               'power, a task that requires '\n",
      "                                               'time-intensive retrospective '\n",
      "                                               'integration. Results: To '\n",
      "                                               'address this challenge, we '\n",
      "                                               'developed MOLGENIS/connect, a '\n",
      "                                               'semi-automatic system to find, '\n",
      "                                               'match and pool data from '\n",
      "                                               'different sources. The system '\n",
      "                                               'shortlists relevant source '\n",
      "                                               'attributes from thousands of '\n",
      "                                               'candidates using '\n",
      "                                               'ontology-based query expansion '\n",
      "                                               'to overcome variations in '\n",
      "                                               'terminology. Then it generates '\n",
      "                                               'algorithms that transform '\n",
      "                                               'source attributes to a common '\n",
      "                                               'target DataSchema. These '\n",
      "                                               'include unit conversion, '\n",
      "                                               'categorical value matching and '\n",
      "                                               'complex conversion patterns '\n",
      "                                               '(e.g. calculation of BMI). In '\n",
      "                                               'comparison to human-experts, '\n",
      "                                               'MOLGENIS/connect was able to '\n",
      "                                               'auto-generate 27% of the '\n",
      "                                               'algorithms perfectly, with an '\n",
      "                                               'additional 46% needing only '\n",
      "                                               'minor editing, representing a '\n",
      "                                               'reduction in the human effort '\n",
      "                                               'and expertise needed to pool '\n",
      "                                               'data. Availability and '\n",
      "                                               'Implementation: Source code, '\n",
      "                                               'binaries and documentation are '\n",
      "                                               'available as open-source under '\n",
      "                                               'LGPLv3 from '\n",
      "                                               'http://github.com/molgenis/molgenis '\n",
      "                                               'and www.molgenis.org/connect. '\n",
      "                                               'Contact: Supplementary '\n",
      "                                               'information: Supplementary '\n",
      "                                               'data are available at '\n",
      "                                               'Bioinformatics online.',\n",
      "                                   'authors': [ {'name': 'Pang C.'},\n",
      "                                                {'name': 'Van Enckevort D.'},\n",
      "                                                {'name': 'De Haan M.'},\n",
      "                                                {'name': 'Kelpin F.'},\n",
      "                                                {'name': 'Jetten J.'},\n",
      "                                                {'name': 'Hendriksen D.'},\n",
      "                                                {'name': 'De Boer T.'},\n",
      "                                                {'name': 'Charbon B.'},\n",
      "                                                {'name': 'Winder E.'},\n",
      "                                                {'name': 'Van Der Velde K.J.'},\n",
      "                                                {'name': 'Doiron D.'},\n",
      "                                                {'name': 'Fortier I.'},\n",
      "                                                {'name': 'Hillege H.'},\n",
      "                                                {'name': 'Swertz M.A.'}],\n",
      "                                   'citationCount': 12,\n",
      "                                   'date': '2016-07-15T00:00:00Z',\n",
      "                                   'journal': 'Bioinformatics',\n",
      "                                   'title': 'MOLGENIS/connect: A system for '\n",
      "                                            'semi-automatic integration of '\n",
      "                                            'heterogeneous phenotype data with '\n",
      "                                            'applications in biobanks'},\n",
      "                     'note': None,\n",
      "                     'pmcid': 'PMC4937195',\n",
      "                     'pmid': '27153686',\n",
      "                     'type': ['Primary'],\n",
      "                     'version': None},\n",
      "                   { 'doi': '10.1186/1471-2105-11-S12-S12',\n",
      "                     'metadata': { 'abstract': 'Background: There is a huge '\n",
      "                                               'demand on bioinformaticians to '\n",
      "                                               'provide their biologists with '\n",
      "                                               'user friendly and scalable '\n",
      "                                               'software infrastructures to '\n",
      "                                               'capture, exchange, and exploit '\n",
      "                                               'the unprecedented amounts of '\n",
      "                                               'new *omics data. We here '\n",
      "                                               'present MOLGENIS, a generic, '\n",
      "                                               'open source, software toolkit '\n",
      "                                               'to quickly produce the bespoke '\n",
      "                                               'MOLecular GENetics Information '\n",
      "                                               'Systems needed.Methods: The '\n",
      "                                               'MOLGENIS toolkit provides '\n",
      "                                               'bioinformaticians with a '\n",
      "                                               'simple language to model '\n",
      "                                               'biological data structures and '\n",
      "                                               'user interfaces. At the push '\n",
      "                                               \"of a button, MOLGENIS' \"\n",
      "                                               'generator suite automatically '\n",
      "                                               'translates these models into a '\n",
      "                                               'feature-rich, ready-to-use web '\n",
      "                                               'application including '\n",
      "                                               'database, user interfaces, '\n",
      "                                               'exchange formats, and '\n",
      "                                               'scriptable interfaces. Each '\n",
      "                                               'generator is a template of '\n",
      "                                               'SQL, JAVA, R, or HTML code '\n",
      "                                               'that would require much effort '\n",
      "                                               'to write by hand. This '\n",
      "                                               \"'model-driven' method ensures \"\n",
      "                                               'reuse of best practices and '\n",
      "                                               'improves quality because the '\n",
      "                                               'modeling language and '\n",
      "                                               'generators are shared between '\n",
      "                                               'all MOLGENIS applications, so '\n",
      "                                               'that errors are found quickly '\n",
      "                                               'and improvements are shared '\n",
      "                                               'easily by a re-generation. A '\n",
      "                                               'plug-in mechanism ensures that '\n",
      "                                               'both the generator suite and '\n",
      "                                               'generated product can be '\n",
      "                                               'customized just as much as '\n",
      "                                               'hand-written software.Results: '\n",
      "                                               'In recent years we have '\n",
      "                                               'successfully evaluated the '\n",
      "                                               'MOLGENIS toolkit for the rapid '\n",
      "                                               'prototyping of many types of '\n",
      "                                               'biomedical applications, '\n",
      "                                               'including next-generation '\n",
      "                                               'sequencing, GWAS, QTL, '\n",
      "                                               'proteomics and biobanking. '\n",
      "                                               'Writing 500 lines of model XML '\n",
      "                                               'typically replaces 15,000 '\n",
      "                                               'lines of hand-written '\n",
      "                                               'programming code, which allows '\n",
      "                                               'for quick adaptation if the '\n",
      "                                               'information system is not yet '\n",
      "                                               \"to the biologist's \"\n",
      "                                               'satisfaction. Each application '\n",
      "                                               'generated with MOLGENIS comes '\n",
      "                                               'with an optimized database '\n",
      "                                               'back-end, user interfaces for '\n",
      "                                               'biologists to manage and '\n",
      "                                               'exploit their data, '\n",
      "                                               'programming interfaces for '\n",
      "                                               'bioinformaticians to script '\n",
      "                                               'analysis tools in R, Java, '\n",
      "                                               'SOAP, REST/JSON and RDF, a '\n",
      "                                               'tab-delimited file format to '\n",
      "                                               'ease upload and exchange of '\n",
      "                                               'data, and detailed technical '\n",
      "                                               'documentation. Existing '\n",
      "                                               'databases can be quickly '\n",
      "                                               'enhanced with MOLGENIS '\n",
      "                                               'generated interfaces using the '\n",
      "                                               \"'ExtractModel' \"\n",
      "                                               'procedure.Conclusions: The '\n",
      "                                               'MOLGENIS toolkit provides '\n",
      "                                               'bioinformaticians with a '\n",
      "                                               'simple model to quickly '\n",
      "                                               'generate flexible web '\n",
      "                                               'platforms for all possible '\n",
      "                                               'genomic, molecular and '\n",
      "                                               'phenotypic experiments with a '\n",
      "                                               'richness of interfaces not '\n",
      "                                               'provided by other tools. All '\n",
      "                                               'the software and manuals are '\n",
      "                                               'available free as LGPLv3 open '\n",
      "                                               'source at '\n",
      "                                               'http://www.molgenis.org. © '\n",
      "                                               '2010 Swertz et al; licensee '\n",
      "                                               'BioMed Central Ltd.',\n",
      "                                   'authors': [ {'name': 'Swertz M.A.'},\n",
      "                                                {'name': 'Dijkstra M.'},\n",
      "                                                {'name': 'Adamusiak T.'},\n",
      "                                                {'name': 'van der Velde J.K.'},\n",
      "                                                {'name': 'Kanterakis A.'},\n",
      "                                                {'name': 'Roos E.T.'},\n",
      "                                                {'name': 'Lops J.'},\n",
      "                                                {'name': 'Thorisson G.A.'},\n",
      "                                                {'name': 'Arends D.'},\n",
      "                                                {'name': 'Byelas G.'},\n",
      "                                                {'name': 'Muilu J.'},\n",
      "                                                {'name': 'Brookes A.J.'},\n",
      "                                                {'name': 'De Brock E.O.'},\n",
      "                                                {'name': 'Jansen R.C.'},\n",
      "                                                {'name': 'Parkinson H.'}],\n",
      "                                   'citationCount': 71,\n",
      "                                   'date': '2010-12-21T00:00:00Z',\n",
      "                                   'journal': 'BMC Bioinformatics',\n",
      "                                   'title': 'The MOLGENIS toolkit: Rapid '\n",
      "                                            'prototyping of biosoftware at the '\n",
      "                                            'push of a button'},\n",
      "                     'note': None,\n",
      "                     'pmcid': None,\n",
      "                     'pmid': None,\n",
      "                     'type': ['Other'],\n",
      "                     'version': None}],\n",
      "  'relation': [],\n",
      "  'toolType': ['Web application'],\n",
      "  'topic': [ { 'term': 'Data management',\n",
      "               'uri': 'http://edamontology.org/topic_3071'},\n",
      "             { 'term': 'Data visualisation',\n",
      "               'uri': 'http://edamontology.org/topic_0092'},\n",
      "             { 'term': 'Rare diseases',\n",
      "               'uri': 'http://edamontology.org/topic_3325'}],\n",
      "  'validated': 1,\n",
      "  'version': ['1']}\n"
     ]
    }
   ],
   "source": [
    "# search for rources from BBMRI collection\n",
    "params = {\n",
    "    #\"q\": [\"multi-omics\"],\n",
    "    \"collectionID\": \"BBMRI\" # \"Rare Disease\", \"RD-connect\"\n",
    "}\n",
    "response = requests.request(\"GET\", url, headers = headers, params = params)\n",
    "\n",
    "# structure of result\n",
    "print(response.json().keys())\n",
    "\n",
    "# number of results\n",
    "print(response.json()[\"count\"])\n",
    "\n",
    "# print 1st result\n",
    "pp.pprint(response.json()[\"list\"][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7db7e5b-07ee-4c2b-a393-f7e2afba1517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44\n",
      "[ 'ProtVar',\n",
      "  'Oligogenic resource for variant analysis (ORVAL)',\n",
      "  'Digenic Diseases Database (DIDA)',\n",
      "  'Variant Combination Pathogenicity Predictor (VarCoPP) 2.0',\n",
      "  'OLIDA: OLIgogenic diseases DAtabase',\n",
      "  'ReMM score',\n",
      "  'ANNOVAR',\n",
      "  'UCSC Genome Browser',\n",
      "  '1000Genomes',\n",
      "  'CADD']\n"
     ]
    }
   ],
   "source": [
    "# resources from rare disease collection\n",
    "params = {\n",
    "    \"q\": \"Genomics\",\n",
    "    \"collectionID\": \"Rare Disease\",\n",
    "    \"accessibility\": \"Open access\"\n",
    "}\n",
    "response = requests.request(\"GET\", url, headers = headers, params = params)\n",
    "print(response.json()[\"count\"])\n",
    "pp.pprint([(record[\"name\"]) for record in response.json()[\"list\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db30a0f9-cf9d-4d80-91cd-af1b45b147bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n"
     ]
    }
   ],
   "source": [
    "# use a general search term\n",
    "params = {\n",
    "    \"q\": \"translational medicine\"\n",
    "}\n",
    "response = requests.request(\"GET\", url, headers = headers, params = params)\n",
    "print(response.json()[\"count\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0811614b-5d0d-4260-a2ee-dfaa81f5fafd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "# query collection EUCAIM, open access only\n",
    "params = {\n",
    "    \"accessibility\": \"Open access\",\n",
    "    \"collectionID\": \"EUCAIM\"\n",
    "}\n",
    "response = requests.request(\"GET\", url, headers = headers, params = params)\n",
    "print(response.json()[\"count\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "586afb68-c554-4d93-9fcf-814d620f09a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['name', 'description', 'homepage', 'biotoolsID', 'biotoolsCURIE', 'version', 'otherID', 'relation', 'function', 'toolType', 'topic', 'operatingSystem', 'language', 'license', 'collectionID', 'maturity', 'cost', 'accessibility', 'elixirPlatform', 'elixirNode', 'elixirCommunity', 'link', 'download', 'documentation', 'publication', 'credit', 'community', 'owner', 'additionDate', 'lastUpdate', 'editPermission', 'validated', 'homepage_status', 'elixir_badge', 'confidence_flag'])\n",
      "[ ('Trace4Harmonization', 'Proprietary'),\n",
      "  ('MITK', 'BSD-3-Clause'),\n",
      "  ('FAIR4Health Data Curation Tool', 'CC-BY-4.0'),\n",
      "  ('Fed-BioMed', 'Apache-2.0'),\n",
      "  ('RadiomicsEnabler', 'Proprietary'),\n",
      "  ('Virtual Imaging Platform', 'CECILL-B'),\n",
      "  ('MuGVRE', 'Apache-2.0')]\n"
     ]
    }
   ],
   "source": [
    "# summarise previous results - record name and license\n",
    "print(response.json()[\"list\"][0].keys())\n",
    "pp.pprint([(record[\"name\"], record[\"license\"]) for record in response.json()[\"list\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a93628c-65b5-46ed-a64d-ace48024f6ed",
   "metadata": {},
   "source": [
    "## Store resources from bio.tools in MOTBX format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b95585a0-bce5-4bec-b5a7-527d97d20790",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query yields 27 items\n",
      "[ ('MINT', 'https://mint.bio.uniroma2.it/'),\n",
      "  ('OMIM', 'http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=OMIM'),\n",
      "  ('ClinVar', 'http://www.ncbi.nlm.nih.gov/clinvar/'),\n",
      "  ('NCBI Resources', 'http://www.ncbi.nlm.nih.gov'),\n",
      "  ('DisGeNET', 'http://www.disgenet.org/'),\n",
      "  ('WikiPathways', 'http://www.wikipathways.org'),\n",
      "  ('Ensembl', 'http://www.ensembl.org/'),\n",
      "  ('dbGAP', 'http://www.ncbi.nlm.nih.gov/gap'),\n",
      "  ('gnomAD', 'http://gnomad.broadinstitute.org/'),\n",
      "  ('ORDO', 'http://www.ebi.ac.uk/ols/ontologies/ordo')]\n",
      "[ ('Orphanet', 'http://www.orpha.net/consor/cgi-bin/index.php'),\n",
      "  ('UMD', 'http://www.umd.be/'),\n",
      "  ('PhenomeCentral', 'https://phenomecentral.org'),\n",
      "  ('EGA', 'https://ega-archive.org/'),\n",
      "  ('Diseasecard', 'http://bioinformatics.ua.pt/diseasecard/'),\n",
      "  ('PharmGKB', 'http://www.pharmgkb.org'),\n",
      "  ('LOVD', 'http://www.lovd.nl/3.0/home'),\n",
      "  ('COEUS', 'http://bioinformatics.ua.pt/coeus/'),\n",
      "  ('RD-Connect Sample Catalogue', 'https://samples.rd-connect.eu/'),\n",
      "  ('DO', 'http://disease-ontology.org/')]\n",
      "[ ('VHLdb', 'http://vhldb.bio.unipd.it/'),\n",
      "  ('COSMIC', 'http://www.sanger.ac.uk/genetics/CGP/cosmic/'),\n",
      "  ('DVD', 'https://trac.nbic.nl/dvd/'),\n",
      "  ('GTEx', 'https://www.gtexportal.org/home/'),\n",
      "  ('Entrez', 'https://www.ncbi.nlm.nih.gov/Entrez/'),\n",
      "  ('DGVa', 'http://www.ebi.ac.uk/dgva'),\n",
      "  ('GeneCards', 'http://www.genecards.org/')]\n"
     ]
    }
   ],
   "source": [
    "# define query parameters\n",
    "params = {\n",
    "    \"collectionID\": \"Rare Diseases\",\n",
    "    \"accessibility\": \"Open access\",\n",
    "    \"toolType\": \"Database portal\",\n",
    "    \"maturity\": \"Mature\",\n",
    "    \"operation\": \"Data retrieval\",  # EDAM operation\n",
    "    \"topic\": \"Data integration and warehousing\",  # EDAM topic\n",
    "    \"page\": 1\n",
    "}\n",
    "\n",
    "# get response from API\n",
    "response = requests.request(\"GET\", url, headers = headers, params = params).json()\n",
    "print(f'Query yields {response[\"count\"]} items')\n",
    "if response[\"count\"] > 0:\n",
    "    pp.pprint([(r[\"name\"], r[\"homepage\"]) for r in response[\"list\"]])\n",
    "\n",
    "all_items = response[\"list\"]  # collect responses\n",
    "# go though all available pages\n",
    "while response[\"next\"] != None:\n",
    "    params[\"page\"] += 1\n",
    "    response = requests.request(\"GET\", url, headers = headers, params = params).json()\n",
    "    pp.pprint([(r[\"name\"], r[\"homepage\"]) for r in response[\"list\"]])\n",
    "    all_items += response[\"list\"]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "17f7639b-c7a4-45c1-bd9d-64f87e6bf65e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MINT https://mint.bio.uniroma2.it/ mint\n",
      "OMIM http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=OMIM omim\n",
      "ClinVar http://www.ncbi.nlm.nih.gov/clinvar/ clinvar\n",
      "NCBI Resources http://www.ncbi.nlm.nih.gov ncbi_resources\n",
      "DisGeNET http://www.disgenet.org/ disgenet\n",
      "WikiPathways http://www.wikipathways.org wikipathways\n",
      "Ensembl http://www.ensembl.org/ ensembl\n",
      "dbGAP http://www.ncbi.nlm.nih.gov/gap dbgap\n",
      "gnomAD http://gnomad.broadinstitute.org/ gnomad\n",
      "ORDO http://www.ebi.ac.uk/ols/ontologies/ordo ordo\n",
      "Orphanet http://www.orpha.net/consor/cgi-bin/index.php orphanet\n",
      "UMD http://www.umd.be/ umd\n",
      "  Resource could not be validated\n",
      "PhenomeCentral https://phenomecentral.org phenomecentral\n",
      "  Resource could not be validated\n",
      "EGA https://ega-archive.org/ ega\n",
      "Diseasecard http://bioinformatics.ua.pt/diseasecard/ diseasecard\n",
      "PharmGKB http://www.pharmgkb.org pharmgkb\n",
      "LOVD http://www.lovd.nl/3.0/home lovd\n",
      "COEUS http://bioinformatics.ua.pt/coeus/ coeus\n",
      "RD-Connect Sample Catalogue https://samples.rd-connect.eu/ rd-connect_sample_catalogue\n",
      "  Resource could not be validated\n",
      "DO http://disease-ontology.org/ disease_ontology\n",
      "VHLdb http://vhldb.bio.unipd.it/ vhldb\n",
      "  Resource could not be validated\n",
      "COSMIC http://www.sanger.ac.uk/genetics/CGP/cosmic/ cosmic\n",
      "DVD https://trac.nbic.nl/dvd/ dvd\n",
      "GTEx https://www.gtexportal.org/home/ gtex\n",
      "Entrez https://www.ncbi.nlm.nih.gov/Entrez/ entrez\n",
      "DGVa http://www.ebi.ac.uk/dgva dgva\n",
      "GeneCards http://www.genecards.org/ genecards\n"
     ]
    }
   ],
   "source": [
    "for item in all_items:\n",
    "    print(item[\"name\"], item[\"homepage\"], item[\"biotoolsID\"])\n",
    "    MOTBX_resource = {\n",
    "        \"resourceID\": item[\"biotoolsCURIE\"],\n",
    "        \"resourceTitle\": f'External resource registered at bio.tools: {item[\"name\"]}',\n",
    "        \"resourceDescription\": item[\"description\"],\n",
    "        \"resourceUrl\": item[\"homepage\"].replace(\"http://\", \"https://\"),\n",
    "        \"resourceCategory\": \"Data Management and Stewardship\",\n",
    "        \"resourceSubcategory\": \"Databases and catalogues\",\n",
    "        \"resourceTags\": [\"database\"],\n",
    "        \"resourceKeywords\": item[\"collectionID\"] + [\n",
    "            o[\"term\"] for i in item[\"function\"] for o in i[\"operation\"]] + [\n",
    "            v[\"term\"] for i in item[\"function\"] for e in i[\"input\"] for k,v in e.items() if isinstance(e, dict) and isinstance(v, dict)] + [\n",
    "            v[\"term\"] for i in item[\"function\"] for e in i[\"output\"] for k,v in e.items() if isinstance(e, dict) and isinstance(v, dict)]\n",
    "    }\n",
    "    if not MOTBX_resource[\"resourceDescription\"].endswith(\".\"):\n",
    "        MOTBX_resource[\"resourceDescription\"] += \".\"\n",
    "    MOTBX_resource[\"resourceDescription\"] += \" Metadata for this resource was programmatically retrieved from elixir's bio.tools registry.\"\n",
    "    # create resource object\n",
    "    resource = motbxschema.MotbxResource(\n",
    "        RESOURCES_DIR.joinpath(f'{item[\"biotoolsID\"]}.yaml'),\n",
    "        resource=MOTBX_resource)\n",
    "    try:\n",
    "        resource.validate(schema)\n",
    "    except Exception:\n",
    "        print(\"  Resource could not be validated\")\n",
    "        continue\n",
    "    # write to YAML file\n",
    "    resource.save()\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11 (motbxtools)",
   "language": "python",
   "name": "motbxtools"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
