{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "831ae188-c669-4885-88ce-6f4ba7718be8",
   "metadata": {},
   "source": [
    "# Summarise MOTBX resources\n",
    "\n",
    "Convert MOTBX resources (YAML files) to CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "94f35a6e-6ad6-4801-b66f-496efc669f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import csv\n",
    "from motbxtools import motbxschema\n",
    "import json\n",
    "import yaml\n",
    "from collections import  defaultdict as ddict\n",
    "import pprint\n",
    "pp = pprint.PrettyPrinter(indent = 2)\n",
    "\n",
    "CWD = Path.cwd()\n",
    "if CWD.name != \"notebooks\":\n",
    "    print(\"Make sure to run this notebook from the 'notebooks' directory.\")\n",
    "\n",
    "MOTBX_DIR = CWD.parent # home directory of this repository\n",
    "\n",
    "with open(MOTBX_DIR.joinpath(\"resources/MOTBX_version.yaml\"), \"r\") as f:\n",
    "    MOTBX_VERSION = yaml.safe_load(f)\n",
    "\n",
    "# path to directory where resources YAML file are saved\n",
    "RESOURCES_DIR = MOTBX_DIR.joinpath(\"resources/curated\")\n",
    "\n",
    "# path to JSON SCHEMA file defining structure of MOTBX resources\n",
    "SCHEMA_JSON = MOTBX_DIR.joinpath(\"schema/motbxschema.json\")\n",
    "\n",
    "TAG_MAPPING = MOTBX_DIR.joinpath(\"tags_keywords.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "5a8e82c5-2fc0-4e13-bf4e-dd7d5e58181b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"$schema\": \"https://json-schema.org/draft/2020-12/schema\",\n",
      "  \"title\": \"MOTBX resource\",\n",
      "  \"description\": \"Schema for resources of the EATRIS Multi-omics Toolbox (MOTBX)\",\n",
      "  \"type\": \"object\",\n",
      "  \"properties\": {\n",
      "    \"resourceID\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"resourceCategory\": {\n",
      "      \"type\": \"string\",\n",
      "      \"enum\": [\n",
      "        \"Epigenomics\",\n",
      "        \"External Quality Assessment\",\n",
      "        \"Genomics\",\n",
      "        \"Internal Quality Control\",\n",
      "        \"Metabolomics\",\n",
      "        \"Omics data management and analysis\",\n",
      "        \"Proteomics\",\n",
      "        \"Transcriptomics\"\n",
      "      ]\n",
      "    },\n",
      "    \"resourceSubcategory\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 4,\n",
      "      \"maxLength\": 30\n",
      "    },\n",
      "    \"resourceTitle\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 15,\n",
      "      \"maxLength\": 160\n",
      "    },\n",
      "    \"resourceDescription\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 50,\n",
      "      \"maxLength\": 2500\n",
      "    },\n",
      "    \"resourceUrl\": {\n",
      "      \"type\": \"string\",\n",
      "      \"format\": \"uri\",\n",
      "      \"pattern\": \"^https://|.pdf$\"\n",
      "    },\n",
      "    \"resourceTags\": {\n",
      "      \"type\": \"array\",\n",
      "      \"items\": {\n",
      "        \"type\": \"string\"\n",
      "      },\n",
      "      \"minItems\": 1\n",
      "    }\n",
      "  },\n",
      "  \"required\": [\n",
      "    \"resourceID\",\n",
      "    \"resourceCategory\",\n",
      "    \"resourceSubcategory\",\n",
      "    \"resourceTitle\",\n",
      "    \"resourceDescription\",\n",
      "    \"resourceUrl\",\n",
      "    \"resourceTags\"\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# load JSON schema from file\n",
    "schema = motbxschema.MotbxSchema(SCHEMA_JSON)\n",
    "\n",
    "# print schema formatted as JSON\n",
    "print(json.dumps(schema.schema, indent = 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "93013f06-ff43-4643-b860-585e0d0026fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['resourceID', 'resourceCategory', 'resourceSubcategory', 'resourceTitle', 'resourceDescription', 'resourceUrl', 'resourceTags']\n"
     ]
    }
   ],
   "source": [
    "# define CSV column names\n",
    "fieldnames = list(schema.schema[\"properties\"].keys())\n",
    "print(fieldnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "2aee6cd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'translational research data': [{'omics': [{'synonyms': ['type of omics',\n",
       "       'omics type',\n",
       "       'experimental data']},\n",
       "     'multi-omics (tag)',\n",
       "     {'genomics (tag)': [{'synonyms': ['genome', 'genomics DNA', 'genotype']},\n",
       "       'whole genome',\n",
       "       {'exome': [{'synonyms': ['whole exome']}]},\n",
       "       'copy number variation',\n",
       "       {'somatic genome variations': [{'synonyms': ['somatic mutations',\n",
       "           'somatic mutation']}]}]},\n",
       "     {'epigenomics (tag)': [{'synonyms': ['epigenetic modifications',\n",
       "         'epigenome']},\n",
       "       {'CpG methylation': [{'synonyms': ['methylation', 'methylome']}]}]},\n",
       "     {'transcriptomics (tag)': [{'synonyms': ['transcriptome',\n",
       "         'gene expression',\n",
       "         'RNA']},\n",
       "       {'mRNA': [{'synonyms': ['messenger RNA']}]},\n",
       "       {'miRNA': [{'synonyms': ['micro RNA', 'microRNA']}]}]},\n",
       "     {'proteomics (tag)': [{'synonyms': ['protein levels',\n",
       "         'proteins',\n",
       "         'proteome']}]},\n",
       "     {'metabolomics (tag)': [{'synonyms': ['metabolite levels',\n",
       "         'metabolite profiling',\n",
       "         'metabolites',\n",
       "         'metabolome']},\n",
       "       {'targeted metabolomics': [{'synonyms': ['targeted metabolite profiling']}]},\n",
       "       {'untargeted metabolomics': [{'synonyms': ['untargeted metabolite profiling']}]},\n",
       "       {'lipidomics': [{'synonyms': ['lipidomic profiling',\n",
       "           'lipid profiling']},\n",
       "         'targeted lipidomics',\n",
       "         'untargeted lipidomics']},\n",
       "       'fluxomics']},\n",
       "     {'microbiomics': [{'synonyms': ['microbiome', 'microbiota']}]},\n",
       "     {'macrobiomics': [{'synonyms': ['macrobiota']}]},\n",
       "     {'glycomics': [{'synonyms': ['glycans',\n",
       "         'glycan profiling',\n",
       "         'glycome',\n",
       "         'glycomic profiling']}]},\n",
       "     {'spatial omics': [{'synonyms': ['omics imaging']}]}]},\n",
       "   {'biomedical imaging': [{'synonyms': ['medical imaging']},\n",
       "     {'radiomics': [{'synonyms': ['radiomic features']}]}]},\n",
       "   {'phenotypic data (tag)': [{'synonyms': ['phenotype',\n",
       "       'clinical data',\n",
       "       'clinical information',\n",
       "       'disease symptoms']}]}]},\n",
       " {'assay type': [{'synonyms': ['measurement type']},\n",
       "   {'mass spectrometry (tag)': [{'high-resolution mass spectrometry': [{'synonyms': ['HR-MS',\n",
       "         'HRMS',\n",
       "         'HR-MS analysis']}]},\n",
       "     'Liquid chromatography (LC)',\n",
       "     'UltraHighPerfomenceLiquidChromatography (UHPLC)',\n",
       "     'HILIC',\n",
       "     {'Reversed phase': ['C18', 'C30']},\n",
       "     'direct infusion - mass spectrometry (DI-MS)']},\n",
       "   {'sequencing (tag)': [{'next generation sequencing': [{'synonyms': ['NGS']}]},\n",
       "     {'DNA sequencing': [{'synonyms': ['DNA sequence analysis',\n",
       "         'DNAseq',\n",
       "         'DNA-seq']},\n",
       "       {'Illumina sequencing': ['HiSeq']}]},\n",
       "     {'RNA sequencing': [{'synonyms': ['RNA-seq']}]},\n",
       "     {'methylation sequencing': [{'synonyms': ['methylation profiling by high throughput sequencing',\n",
       "         'methyl-seq']},\n",
       "       {'EM-seq': [{'synonyms': ['NEBNext Enzymatic Methyl-seq',\n",
       "           'Enzymatic Methyl-seq Kit',\n",
       "           'EM-seq NGS']}]}]}]},\n",
       "   'array (tag)',\n",
       "   {'PCR (tag)': [{'synonyms': ['polymerase chain reaction']},\n",
       "     {'RT-PCR': [{'synonyms': ['reverse transcription PCR']},\n",
       "       'microRNA profiling by RT-PCR',\n",
       "       {'qRT-PCR': [{'synonyms': ['quantitative RT-PCR', 'RT-qPCR']}]}]}]}]},\n",
       " {'biospecimen type': [{'synonyms': ['sample type',\n",
       "     'specimen type',\n",
       "     'type of biospecimen']},\n",
       "   {'single-cell (tag)': [{'synonyms': ['single cell sample',\n",
       "       'single cell specimen',\n",
       "       'single cell analysis']}]},\n",
       "   {'biological fluid (tag)': [{'synonyms': ['biofluid', 'biofluids']},\n",
       "     {'blood': [{'synonyms': ['blood sample', 'blood specimen']},\n",
       "       'serum',\n",
       "       'plasma']},\n",
       "     'saliva']},\n",
       "   'tissue (tag)']},\n",
       " {'molecular entity': [{'synonyms': ['material entity', 'sample']},\n",
       "   {'organic polymer': [{'DNA (tag)': [{'synonyms': ['deoxyribonucleic acid']},\n",
       "       'unmethylated lambda DNA']},\n",
       "     {'RNA': [{'synonyms': ['ribonucleic acid']},\n",
       "       'mRNA (tag)',\n",
       "       'miRNA (tag)']},\n",
       "     'protein (tag)']},\n",
       "   {'small molecule': ['amino acid (tag)',\n",
       "     {'metabolite (tag)': ['polar metabolites', 'non-polar metabolites']},\n",
       "     'nucleic acid (tag)',\n",
       "     'lipid (tag)']}]},\n",
       " {'type of resource': [{'proficiency testing (tag)': [{'synonyms': ['PT',\n",
       "       'proficiency testing scheme',\n",
       "       'EQA',\n",
       "       'PT scheme',\n",
       "       'ring trial',\n",
       "       'inter-laboratory tests']}]},\n",
       "   {'translational research use case': [{'synonyms': ['translational use case',\n",
       "       'transaltional example',\n",
       "       'translational research',\n",
       "       'translational medicine',\n",
       "       'translational science']},\n",
       "     {'preclinical study (tag)': [{'synonyms': ['basic research',\n",
       "         'discovery science',\n",
       "         'preclinical testing',\n",
       "         'basic research study']}]},\n",
       "     {'clinical study (tag)': [{'synonyms': ['clinical testing']},\n",
       "       {'phase I clinical trial': [{'synonyms': ['proof of concept',\n",
       "           'translation to humans',\n",
       "           'first-in-human']}]},\n",
       "       {'phase II or III clinical trial': [{'synonyms': ['translation to patients']}]}]}]},\n",
       "   {'scientific publication (tag)': [{'synonyms': ['paper']},\n",
       "     'original research article',\n",
       "     'review article',\n",
       "     {'short communication': [{'synonyms': ['short report',\n",
       "         'letter',\n",
       "         'editorial']}]},\n",
       "     'case study',\n",
       "     'method article',\n",
       "     'meta-analysis']},\n",
       "   {'policy (tag)': [{'synonyms': ['regulation']}]},\n",
       "   {'training': [{'tutorial': [{'synonyms': ['tutorials']}]}, 'template']},\n",
       "   {'guideline (tag)': [{'synonyms': ['recommendations',\n",
       "       'guidelines',\n",
       "       'best practices',\n",
       "       'best practice guidance',\n",
       "       'framework',\n",
       "       'good practices']},\n",
       "     {'FAIR principles': [{'synonyms': ['FAIR guiding principles',\n",
       "         'FAIR']}]}]},\n",
       "   {'software application (tag)': [{'synonyms': ['tool',\n",
       "       'software',\n",
       "       'application']},\n",
       "     {'web application': ['analysis platform']},\n",
       "     'desktop application',\n",
       "     'command-line tool',\n",
       "     {'computational workflow': [{'synonyms': ['pipeline',\n",
       "         'bioinformatics workflow',\n",
       "         'analysis workflow',\n",
       "         'analysis workflows']}]},\n",
       "     'script',\n",
       "     'notebook',\n",
       "     'source code']},\n",
       "   {'reference material (tag)': [{'synonyms': ['quality control material']},\n",
       "     {'certified reference material': [{'synonyms': ['reference standard']}]}]},\n",
       "   'reference data set (tag)',\n",
       "   {'standard (tag)': [{'synonyms': ['specification', 'reference']}]},\n",
       "   {'standardisation process': [{'quality standard': [{'ISO standard (tag)': [{'synonyms': ['international standard']}]},\n",
       "       {'internal quality control (IQC) (tag)': [{'synonyms': ['IQC',\n",
       "           'internal quality control',\n",
       "           'in-house quality control',\n",
       "           'in-house']}]}]},\n",
       "     {'data standard (tag)': [{'vocabulary': [{'synonyms': ['terminology']}]},\n",
       "       'data schema',\n",
       "       'data format']},\n",
       "     {'metadata standard': [{'synonyms': ['metadata', 'metadata standards']},\n",
       "       'phenopackets',\n",
       "       {'ISA': [{'synonyms': ['investigation-study-assay']}]}]}]},\n",
       "   {'protocol (tag)': [{'synonyms': ['SOP',\n",
       "       'standard operating procedure',\n",
       "       'procedure',\n",
       "       'step-by-step instructions']}]},\n",
       "   {'organisation (tag)': [{'synonyms': ['organization']}, 'EATRIS']},\n",
       "   {'project': [{'EATRIS flagship project': [{'synonyms': ['EATRIS project',\n",
       "         'flagship project']},\n",
       "       'EATRIS-Plus (tag)']}]},\n",
       "   {'registry (tag)': [{'synonyms': ['repository']},\n",
       "     {'computational workflow registry': [{'synonyms': ['workflow registry']}]},\n",
       "     {'software registry': [{'synonyms': ['bioinformatics software registry',\n",
       "         'tool registry']}]},\n",
       "     'services registry',\n",
       "     'standards registry',\n",
       "     'database registry',\n",
       "     {'data catalogue': [{'synonyms': ['metadata repository',\n",
       "         'data registry',\n",
       "         'data catalog']}]}]},\n",
       "   {'database (tag)': [{'synonyms': ['database portal',\n",
       "       'data repository',\n",
       "       'data archive',\n",
       "       'data archives']}]},\n",
       "   {'knowledgebase (tag)': [{'synonyms': ['resource collection']}]}]}]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(TAG_MAPPING, \"r\") as file:\n",
    "    tag_mapping = yaml.safe_load(file)\n",
    "\n",
    "tag_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41, 344)"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def unnest(nested, pre=None):\n",
    "    pre = pre[:] if pre else []\n",
    "    if isinstance(nested, dict):\n",
    "        for key, value in nested.items():\n",
    "            if isinstance(value, dict):\n",
    "                for d in unnest(value, pre + [key]):\n",
    "                    yield d\n",
    "            elif isinstance(value, list):\n",
    "                for v in value:\n",
    "                    for d in unnest(v, pre + [key]):\n",
    "                        yield d\n",
    "            else:\n",
    "                yield pre + [key, value]\n",
    "    elif isinstance(nested, list):\n",
    "        for value in nested:\n",
    "            if isinstance(value, dict):\n",
    "                for d in unnest(value, pre):\n",
    "                    yield d\n",
    "            elif isinstance(value, list):\n",
    "                for v in value:\n",
    "                    for d in unnest(v, pre):\n",
    "                        yield d\n",
    "            else:\n",
    "                yield pre + [key, value]\n",
    "    else:\n",
    "        yield pre + [nested]\n",
    "\n",
    "tags = set()\n",
    "term2parents = ddict(set)\n",
    "term2synonyms = ddict(set)\n",
    "for l in unnest(tag_mapping):\n",
    "    _tags = set([i for i in l if i[-5:] == \"(tag)\"])\n",
    "    if len(_tags) > 1:\n",
    "        print(_tags)\n",
    "    tags |= _tags\n",
    "    if \"synonyms\" in l:\n",
    "        term2synonyms[l[-3]].add(l[-1])\n",
    "    for idx, i in enumerate(l):\n",
    "        if i == \"synonyms\":\n",
    "            break\n",
    "        term2parents[i] |= set(l[:idx])\n",
    "\n",
    "synonym2terms = ddict(set)\n",
    "for k, v in term2synonyms.items():\n",
    "    for i in v:\n",
    "        synonym2terms[i].add(k)\n",
    "\n",
    "term2keywords = ddict(set)\n",
    "for k, v in term2parents.items():\n",
    "    term2keywords[k] |= v.union(*[term2synonyms[i] for i in v])\n",
    "\n",
    "len(tags), len(term2keywords) + len([i for j in term2synonyms.values() for i in j]) + 62"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "be099a71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID0001.yaml\n",
      "ID0002.yaml\n",
      "documentation !\n",
      "ID0003.yaml\n",
      "ID0004.yaml\n",
      "miRNA-seq !\n",
      "nucleic acid quantitation !\n",
      "ID0005.yaml\n",
      "nucleic acid quantitation !\n",
      "ID0006.yaml\n",
      "WGS !\n",
      "ID0007.yaml\n",
      "ID0008.yaml\n",
      "reference dataset !\n",
      "ID0009.yaml\n",
      "methyl-seq kit !\n",
      "ID0010.yaml\n",
      "reference dataset !\n",
      "ID0011.yaml\n",
      "EMQN !\n",
      "certification !\n",
      "external quality assessment (EQA) !\n",
      "quality !\n",
      "ID0012.yaml\n",
      "EQIPD !\n",
      "certification !\n",
      "external quality assessment (EQA) !\n",
      "quality !\n",
      "toolbox !\n",
      "ID0013.yaml\n",
      "ECCQ !\n",
      "certification !\n",
      "external quality assessment (EQA) !\n",
      "quality !\n",
      "ID0014.yaml\n",
      "external quality assessment (EQA) !\n",
      "proficiency testing programme !\n",
      "ID0015.yaml\n",
      "WGS !\n",
      "workflow !\n",
      "ID0016.yaml\n",
      "workflow !\n",
      "ID0017.yaml\n",
      "SRiC !\n",
      "publication !\n",
      "ID0018.yaml\n",
      "toolkit !\n",
      "ID0019.yaml\n",
      "publication !\n",
      "ID0020.yaml\n",
      "data processing !\n",
      "library preparation !\n",
      "publication !\n",
      "ID0021.yaml\n",
      "publication !\n",
      "ID0022.yaml\n",
      "ID0023.yaml\n",
      "publication !\n",
      "ID0024.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0025.yaml\n",
      "Hi-seq !\n",
      "data quality !\n",
      "publication !\n",
      "ID0026.yaml\n",
      "BWAmeth !\n",
      "NFcore !\n",
      "alignment !\n",
      "bismark !\n",
      "data processing !\n",
      "mapping !\n",
      "ID0027.yaml\n",
      "data processing !\n",
      "miRNAs !\n",
      "normalization !\n",
      "publication !\n",
      "ID0028.yaml\n",
      "MIQE !\n",
      "miRNAs !\n",
      "normalization !\n",
      "standards !\n",
      "ID0029.yaml\n",
      "MIQE !\n",
      "publication !\n",
      "qPCR !\n",
      "reproducibility !\n",
      "ID0030.yaml\n",
      "MIQE !\n",
      "publication !\n",
      "qPCR !\n",
      "ID0031.yaml\n",
      "MIQE !\n",
      "publication !\n",
      "qPCR !\n",
      "ID0032.yaml\n",
      "EATRIS-Plus project !\n",
      "sample processing protocol !\n",
      "ID0033.yaml\n",
      "EATRIS-Plus project !\n",
      "sample processing protocol !\n",
      "ID0034.yaml\n",
      "miRNAs !\n",
      "publication !\n",
      "qPCR !\n",
      "quality control !\n",
      "standards !\n",
      "ID0035.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0036.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0037.yaml\n",
      "EATRIS-Plus project !\n",
      "extraction !\n",
      "isolation !\n",
      "ID0038.yaml\n",
      "EATRIS-Plus project !\n",
      "Retrotranscription !\n",
      "cDNA synthesis !\n",
      "ID0039.yaml\n",
      "EATRIS-Plus project !\n",
      "qPCR arrays !\n",
      "qPCR panels !\n",
      "ID0040.yaml\n",
      "data processing !\n",
      "miRNAs !\n",
      "normalization !\n",
      "publication !\n",
      "qPCR !\n",
      "ID0041.yaml\n",
      "miRNAs !\n",
      "microRNA array profiling !\n",
      "preclinical use case !\n",
      "publication !\n",
      "qPCR !\n",
      "ID0042.yaml\n",
      "biomarkers !\n",
      "miRNAs !\n",
      "microRNA array profiling !\n",
      "preclinical use case !\n",
      "publication !\n",
      "qPCR !\n",
      "ID0043.yaml\n",
      "biomarkers !\n",
      "clinical validation !\n",
      "miRNAs !\n",
      "publication !\n",
      "qPCR !\n",
      "ID0044.yaml\n",
      "publication !\n",
      "quality control !\n",
      "ID0045.yaml\n",
      "publication !\n",
      "quality assessment !\n",
      "quality control !\n",
      "workflow !\n",
      "ID0046.yaml\n",
      "biomarker !\n",
      "publication !\n",
      "quality control !\n",
      "sample collection !\n",
      "study design !\n",
      "ID0047.yaml\n",
      "publication !\n",
      "study design !\n",
      "ID0048.yaml\n",
      "publication !\n",
      "sample collection !\n",
      "study design !\n",
      "ID0049.yaml\n",
      "publication !\n",
      "ID0050.yaml\n",
      "clinical guidelines !\n",
      "publication !\n",
      "samples !\n",
      "ID0051.yaml\n",
      "publication !\n",
      "ID0052.yaml\n",
      "EATRIS-Plus project !\n",
      "sample preparation !\n",
      "ID0053.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0054.yaml\n",
      "EATRIS-Plus project !\n",
      "data search !\n",
      "ID0055.yaml\n",
      "data processing !\n",
      "toolbox !\n",
      "ID0056.yaml\n",
      "data search !\n",
      "ID0057.yaml\n",
      "data analysis !\n",
      "visualisation tool !\n",
      "ID0058.yaml\n",
      "publication !\n",
      "quality assurance !\n",
      "quality control !\n",
      "reference materials !\n",
      "ID0059.yaml\n",
      "publication !\n",
      "quality assurance !\n",
      "quality control !\n",
      "reference materials !\n",
      "ID0060.yaml\n",
      "publication !\n",
      "quality assurance !\n",
      "quality control !\n",
      "reference materials !\n",
      "ID0061.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0062.yaml\n",
      "EATRIS-Plus project !\n",
      "small molecules !\n",
      "ID0063.yaml\n",
      "MZmine !\n",
      "data processing !\n",
      "ID0064.yaml\n",
      "biomarkers !\n",
      "publication !\n",
      "whole-exome sequencing !\n",
      "ID0065.yaml\n",
      "ID0066.yaml\n",
      "ID0067.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0068.yaml\n",
      "ID0069.yaml\n",
      "data standards !\n",
      "ID0070.yaml\n",
      "ID0071.yaml\n",
      "study management !\n",
      "ID0072.yaml\n",
      "study management !\n",
      "ID0073.yaml\n",
      "study management !\n",
      "ID0074.yaml\n",
      "ID0075.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0076.yaml\n",
      "ID0077.yaml\n",
      "bioinformatics !\n",
      "ID0078.yaml\n",
      "ID0079.yaml\n",
      "ID0080.yaml\n",
      "EATRIS-Plus project !\n",
      "ID0081.yaml\n",
      "reference dataset !\n"
     ]
    }
   ],
   "source": [
    "# iterate through resources\n",
    "for root, dirs, files in os.walk(RESOURCES_DIR):\n",
    "    for name in files:\n",
    "        if not name.endswith(\".yaml\"):\n",
    "            continue\n",
    "        print(name)\n",
    "\n",
    "        # load test resource and validate\n",
    "        resource = motbxschema.MotbxResource(os.path.join(root, name))\n",
    "        #resource.validate(schema) don't validate in first version\n",
    "        resource_tags = set()\n",
    "        resource_keywords = set()\n",
    "        for i in resource.resource[\"resourceTags\"]:\n",
    "            i = i.lower()\n",
    "            if f'{i} (tag)' in tags:\n",
    "                # add old tag to new tags\n",
    "                resource_tags.add(i)\n",
    "                # add synonyms defined for this tag as keywords\n",
    "                resource_keywords |= set(term2keywords[f'{i} (tag)'])\n",
    "            elif i in term2parents:\n",
    "                resource_keywords.add(i)\n",
    "                resource_keywords |= set(term2keywords[i])\n",
    "            else:\n",
    "                #print(i)\n",
    "                if i in synonym2terms.keys():\n",
    "                    #print(synonym2terms[i])\n",
    "                    resource_keywords.add(i)\n",
    "                    for t in synonym2terms[i]:\n",
    "                        if t in tags:\n",
    "                            resource_tags.add(t[:-6])\n",
    "                            resource_keywords |= set(term2synonyms[t])\n",
    "                        else:\n",
    "                            resource_keywords.add(t)\n",
    "                else:\n",
    "                    print(i, \"!\")\n",
    "                    resource_keywords.add(i)\n",
    "        #print(resource_tags)\n",
    "        #pp.pprint(sorted(resource_keywords))\n",
    "        #break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11 (motbxtools)",
   "language": "python",
   "name": "motbxtools"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
