{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "831ae188-c669-4885-88ce-6f4ba7718be8",
   "metadata": {},
   "source": [
    "# Summarise MOTBX resources\n",
    "\n",
    "Convert MOTBX resources (YAML files) to CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "94f35a6e-6ad6-4801-b66f-496efc669f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import csv\n",
    "from motbxtools import motbxschema\n",
    "import json\n",
    "import yaml\n",
    "from collections import  defaultdict as ddict\n",
    "import pprint\n",
    "pp = pprint.PrettyPrinter(indent = 2)\n",
    "\n",
    "CWD = Path.cwd()\n",
    "if CWD.name != \"notebooks\":\n",
    "    print(\"Make sure to run this notebook from the 'notebooks' directory.\")\n",
    "\n",
    "MOTBX_DIR = CWD.parent # home directory of this repository\n",
    "\n",
    "with open(MOTBX_DIR.joinpath(\"resources/MOTBX_version.yaml\"), \"r\") as f:\n",
    "    MOTBX_VERSION = yaml.safe_load(f)\n",
    "\n",
    "# path to directory where resources YAML file are saved\n",
    "RESOURCES_DIR = MOTBX_DIR.joinpath(\"resources/curated\")\n",
    "\n",
    "# path to JSON SCHEMA file defining structure of MOTBX resources\n",
    "SCHEMA_JSON = MOTBX_DIR.joinpath(\"schema/motbxschema.json\")\n",
    "\n",
    "TAG_MAPPING = MOTBX_DIR.joinpath(\"schema/tags_keywords.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5a8e82c5-2fc0-4e13-bf4e-dd7d5e58181b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"$schema\": \"https://json-schema.org/draft/2020-12/schema\",\n",
      "  \"title\": \"MOTBX resource\",\n",
      "  \"description\": \"Schema for resources of the EATRIS Multi-omics Toolbox (MOTBX)\",\n",
      "  \"type\": \"object\",\n",
      "  \"properties\": {\n",
      "    \"resourceID\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"resourceCategory\": {\n",
      "      \"type\": \"string\",\n",
      "      \"enum\": [\n",
      "        \"Epigenomics\",\n",
      "        \"External Quality Assessment\",\n",
      "        \"Genomics\",\n",
      "        \"Internal Quality Control\",\n",
      "        \"Metabolomics\",\n",
      "        \"Omics data management and analysis\",\n",
      "        \"Proteomics\",\n",
      "        \"Transcriptomics\"\n",
      "      ]\n",
      "    },\n",
      "    \"resourceSubcategory\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 4,\n",
      "      \"maxLength\": 30\n",
      "    },\n",
      "    \"resourceTitle\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 15,\n",
      "      \"maxLength\": 160\n",
      "    },\n",
      "    \"resourceDescription\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 50,\n",
      "      \"maxLength\": 2500\n",
      "    },\n",
      "    \"resourceUrl\": {\n",
      "      \"type\": \"string\",\n",
      "      \"format\": \"uri\",\n",
      "      \"pattern\": \"^https://|.pdf$\"\n",
      "    },\n",
      "    \"resourceTags\": {\n",
      "      \"type\": \"array\",\n",
      "      \"items\": {\n",
      "        \"type\": \"string\"\n",
      "      },\n",
      "      \"minItems\": 1\n",
      "    }\n",
      "  },\n",
      "  \"required\": [\n",
      "    \"resourceID\",\n",
      "    \"resourceCategory\",\n",
      "    \"resourceSubcategory\",\n",
      "    \"resourceTitle\",\n",
      "    \"resourceDescription\",\n",
      "    \"resourceUrl\",\n",
      "    \"resourceTags\"\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# load JSON schema from file\n",
    "schema = motbxschema.MotbxSchema(SCHEMA_JSON)\n",
    "\n",
    "# print schema formatted as JSON\n",
    "print(json.dumps(schema.schema, indent = 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "93013f06-ff43-4643-b860-585e0d0026fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['resourceID', 'resourceCategory', 'resourceSubcategory', 'resourceTitle', 'resourceDescription', 'resourceUrl', 'resourceTags']\n"
     ]
    }
   ],
   "source": [
    "# define CSV column names\n",
    "fieldnames = list(schema.schema[\"properties\"].keys())\n",
    "print(fieldnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2aee6cd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'translational research data': [{'omics': [{'synonyms': ['type of omics',\n",
       "       'omics type',\n",
       "       'experimental data']},\n",
       "     'multi-omics (tag)',\n",
       "     {'genomics (tag)': [{'synonyms': ['genome', 'genomics DNA', 'genotype']},\n",
       "       {'whole genome': [{'synonyms': ['WGS']}]},\n",
       "       {'exome': [{'synonyms': ['whole exome', 'whole-exome sequencing']}]},\n",
       "       'copy number variation',\n",
       "       {'somatic genome variations': [{'synonyms': ['somatic mutations',\n",
       "           'somatic mutation']}]}]},\n",
       "     {'epigenomics (tag)': [{'synonyms': ['epigenetic modifications',\n",
       "         'epigenome']},\n",
       "       {'CpG methylation': [{'synonyms': ['methylation', 'methylome']}]}]},\n",
       "     {'transcriptomics (tag)': [{'synonyms': ['transcriptome',\n",
       "         'gene expression',\n",
       "         'RNA']},\n",
       "       {'mRNA-seq': [{'synonyms': ['mRNA',\n",
       "           'mRNA sequencing',\n",
       "           'mRNA Seq',\n",
       "           'messenger RNA']}]},\n",
       "       {'miRNA-seq': [{'synonyms': ['miRNA',\n",
       "           'miRNA sequencing',\n",
       "           'miRNA Seq',\n",
       "           'micro RNA',\n",
       "           'microRNA',\n",
       "           'miRNAs']}]}]},\n",
       "     {'proteomics (tag)': [{'synonyms': ['protein levels',\n",
       "         'proteins',\n",
       "         'proteome']}]},\n",
       "     {'metabolomics (tag)': [{'synonyms': ['metabolite levels',\n",
       "         'metabolite profiling',\n",
       "         'metabolites',\n",
       "         'metabolome']},\n",
       "       {'targeted metabolomics': [{'synonyms': ['targeted metabolite profiling']}]},\n",
       "       {'untargeted metabolomics': [{'synonyms': ['untargeted metabolite profiling']}]},\n",
       "       {'lipidomics': [{'synonyms': ['lipidomic profiling',\n",
       "           'lipid profiling']},\n",
       "         'targeted lipidomics',\n",
       "         'untargeted lipidomics']},\n",
       "       'fluxomics']},\n",
       "     {'microbiomics': [{'synonyms': ['microbiome', 'microbiota']}]},\n",
       "     {'macrobiomics': [{'synonyms': ['macrobiota']}]},\n",
       "     {'glycomics': [{'synonyms': ['glycans',\n",
       "         'glycan profiling',\n",
       "         'glycome',\n",
       "         'glycomic profiling']}]},\n",
       "     {'spatial omics': [{'synonyms': ['omics imaging']}]}]},\n",
       "   {'biomedical imaging': [{'synonyms': ['medical imaging']},\n",
       "     {'radiomics': [{'synonyms': ['radiomic features']}]}]},\n",
       "   {'phenotypic data (tag)': [{'synonyms': ['phenotype',\n",
       "       'clinical data',\n",
       "       'clinical information',\n",
       "       'disease symptoms']}]},\n",
       "   {'biomarker (tag)': [{'synonyms': ['biomarkers']}, 'digital biomarker']}]},\n",
       " {'assay type': [{'synonyms': ['measurement type']},\n",
       "   {'mass spectrometry (tag)': [{'high-resolution mass spectrometry': [{'synonyms': ['HR-MS',\n",
       "         'HRMS',\n",
       "         'HR-MS analysis']}]},\n",
       "     'Liquid chromatography (LC)',\n",
       "     'UltraHighPerfomenceLiquidChromatography (UHPLC)',\n",
       "     'HILIC',\n",
       "     {'Reversed phase': ['C18', 'C30']},\n",
       "     'direct infusion - mass spectrometry (DI-MS)']},\n",
       "   {'sequencing (tag)': [{'next generation sequencing': [{'synonyms': ['NGS']}]},\n",
       "     {'DNA sequencing': [{'synonyms': ['DNA sequence analysis',\n",
       "         'DNAseq',\n",
       "         'DNA-seq']},\n",
       "       {'Illumina sequencing': [{'HiSeq': [{'synonyms': ['Hi-seq']}]}]}]},\n",
       "     {'RNA sequencing': [{'synonyms': ['RNA-seq']}]},\n",
       "     {'methylation sequencing': [{'synonyms': ['methylation profiling by high throughput sequencing',\n",
       "         'methyl-seq']},\n",
       "       {'EM-seq': [{'synonyms': ['NEBNext Enzymatic Methyl-seq',\n",
       "           'Enzymatic Methyl-seq Kit',\n",
       "           'EM-seq NGS']}]}]}]},\n",
       "   {'array (tag)': ['microRNA array profiling']},\n",
       "   {'PCR (tag)': [{'synonyms': ['polymerase chain reaction']},\n",
       "     {'qPCR': [{'qPCR array': [{'synonyms': ['qPCR arrays',\n",
       "           'qPCR panels']}]}]},\n",
       "     {'RT-PCR': [{'synonyms': ['reverse transcription PCR']},\n",
       "       'microRNA profiling by RT-PCR',\n",
       "       {'qRT-PCR': [{'synonyms': ['quantitative RT-PCR', 'RT-qPCR']}]}]}]}]},\n",
       " {'biospecimen type': [{'synonyms': ['sample type',\n",
       "     'specimen type',\n",
       "     'type of biospecimen']},\n",
       "   {'single-cell (tag)': [{'synonyms': ['single cell sample',\n",
       "       'single cell specimen',\n",
       "       'single cell analysis']}]},\n",
       "   {'biological fluid (tag)': [{'synonyms': ['biofluid', 'biofluids']},\n",
       "     {'blood': [{'synonyms': ['blood sample', 'blood specimen']},\n",
       "       'serum',\n",
       "       'plasma']},\n",
       "     'saliva']},\n",
       "   'tissue (tag)']},\n",
       " {'molecular entity': [{'synonyms': ['material entity', 'sample']},\n",
       "   {'organic polymer': [{'DNA (tag)': [{'synonyms': ['deoxyribonucleic acid']},\n",
       "       'unmethylated lambda DNA']},\n",
       "     {'RNA': [{'synonyms': ['ribonucleic acid']},\n",
       "       'mRNA (tag)',\n",
       "       'miRNA (tag)']},\n",
       "     'protein (tag)']},\n",
       "   {'small molecule': [{'synonyms': ['small molecules']},\n",
       "     'amino acid (tag)',\n",
       "     {'metabolite (tag)': ['polar metabolites', 'non-polar metabolites']},\n",
       "     'nucleic acid (tag)',\n",
       "     'lipid (tag)']}]},\n",
       " {'type of resource': [{'quality': ['reproducibility',\n",
       "     'data quality',\n",
       "     {'quality assessment': [{'synonyms': ['quality assurance']},\n",
       "       {'quality certification': [{'ECCQ': [{'synonyms': ['EATRIS Certificate of Commitment to Quality']}]},\n",
       "         {'proficiency testing (tag)': [{'synonyms': ['PT',\n",
       "             'proficiency testing scheme',\n",
       "             'proficiency testing programme',\n",
       "             'EQA',\n",
       "             'external quality assessment (EQA)',\n",
       "             'PT scheme',\n",
       "             'ring trial',\n",
       "             'inter-laboratory tests']}]}]}]},\n",
       "     {'quality control': [{'internal quality control (IQC) (tag)': [{'synonyms': ['IQC',\n",
       "           'internal quality control',\n",
       "           'in-house quality control',\n",
       "           'in-house']}]}]}]},\n",
       "   {'translational research use case': [{'synonyms': ['translational use case',\n",
       "       'transaltional example',\n",
       "       'translational research',\n",
       "       'translational medicine',\n",
       "       'translational science']},\n",
       "     {'preclinical study (tag)': [{'synonyms': ['preclinical use case',\n",
       "         'basic research',\n",
       "         'discovery science',\n",
       "         'preclinical testing',\n",
       "         'basic research study']}]},\n",
       "     {'clinical study (tag)': [{'synonyms': ['clinical testing']},\n",
       "       {'phase I clinical trial': [{'synonyms': ['proof of concept',\n",
       "           'translation to humans',\n",
       "           'first-in-human']}]},\n",
       "       {'phase II or III clinical trial': [{'synonyms': ['translation to patients']}]}]}]},\n",
       "   {'scientific publication (tag)': [{'synonyms': ['paper', 'publication']},\n",
       "     'original research article',\n",
       "     'review article',\n",
       "     {'short communication': [{'synonyms': ['short report',\n",
       "         'letter',\n",
       "         'editorial']}]},\n",
       "     'case study',\n",
       "     'method article',\n",
       "     'meta-analysis']},\n",
       "   {'policy (tag)': [{'synonyms': ['regulation']}]},\n",
       "   {'training': [{'tutorial': [{'synonyms': ['tutorials']}]}, 'template']},\n",
       "   {'guideline (tag)': [{'synonyms': ['recommendations',\n",
       "       'guidelines',\n",
       "       'best practices',\n",
       "       'best practice guidance',\n",
       "       'framework',\n",
       "       'good practices']},\n",
       "     'clinical guidelines',\n",
       "     {'FAIR principles': [{'synonyms': ['FAIR guiding principles',\n",
       "         'FAIR']}]}]},\n",
       "   {'software application (tag)': [{'synonyms': ['tool',\n",
       "       'software',\n",
       "       'application']},\n",
       "     {'web application': ['analysis platform']},\n",
       "     'desktop application',\n",
       "     'command-line tool',\n",
       "     {'computational workflow': [{'synonyms': ['pipeline',\n",
       "         'bioinformatics workflow',\n",
       "         'analysis workflow',\n",
       "         'analysis workflows']}]},\n",
       "     'script',\n",
       "     'notebook',\n",
       "     'source code',\n",
       "     {'bioinformatics tool': ['MZmine', 'bismark']},\n",
       "     {'workflow management system': [{'nextflow': [{'nfcore': [{'synonyms': ['NFcore']}]}]}]}]},\n",
       "   {'data processing': [{'synonyms': ['data analysis']},\n",
       "     'data search',\n",
       "     {'normalisation': [{'synonyms': ['normalization']}]},\n",
       "     'alignment',\n",
       "     'mapping',\n",
       "     {'data visualisation': [{'synonyms': ['data visualization',\n",
       "         'visualisation',\n",
       "         'visualization']}]}]},\n",
       "   {'reference material (tag)': [{'synonyms': ['quality control material',\n",
       "       'reference materials']},\n",
       "     {'certified reference material': [{'synonyms': ['reference standard']}]}]},\n",
       "   {'reference data set (tag)': [{'synonyms': ['reference dataset']}]},\n",
       "   {'standard': [{'synonyms': ['standards', 'specification', 'reference']},\n",
       "     {'standardisation process': [{'quality standard': [{'ISO standard (tag)': [{'synonyms': ['international standard']}]}]}]},\n",
       "     {'minimum information standard (tag)': [{'MIQE': [{'synonyms': ['minimum information for publication of quantitative real-time PCR experiments']}]}]},\n",
       "     {'data standard (tag)': [{'synonyms': ['data standards']},\n",
       "       {'vocabulary': [{'synonyms': ['terminology']}]},\n",
       "       'data schema',\n",
       "       'data format']},\n",
       "     {'metadata standard (tag)': [{'synonyms': ['metadata',\n",
       "         'metadata standards']},\n",
       "       'phenopackets',\n",
       "       {'ISA': [{'synonyms': ['investigation-study-assay']}]}]}]},\n",
       "   {'protocol (tag)': [{'synonyms': ['SOP',\n",
       "       'standard operating procedure',\n",
       "       'procedure',\n",
       "       'step-by-step instructions']},\n",
       "     {'sample preparation protocol': [{'synonyms': ['samples preparation',\n",
       "         'sample preparation',\n",
       "         'sample collection',\n",
       "         'sample processing',\n",
       "         'sample processing protocol']},\n",
       "       {'extraction': [{'synonyms': ['isolation']}]},\n",
       "       {'cDNA synthesis': [{'synonyms': ['retrotranscription',\n",
       "           'Retrotranscription']}]},\n",
       "       'library preparation',\n",
       "       'nucleic acid quantification']}]},\n",
       "   {'organisation (tag)': [{'synonyms': ['organization']}, 'EATRIS']},\n",
       "   {'project': [{'EATRIS flagship project': [{'synonyms': ['EATRIS project',\n",
       "         'flagship project']},\n",
       "       {'EATRIS-Plus (tag)': [{'synonyms': ['EATRIS-Plus project']}]}]},\n",
       "     {'EQIPD': [{'synonyms': ['European Quality In Preclinical Data']}]}]},\n",
       "   {'registry (tag)': [{'synonyms': ['repository']},\n",
       "     {'computational workflow registry': [{'synonyms': ['workflow registry']}]},\n",
       "     {'software registry': [{'synonyms': ['bioinformatics software registry',\n",
       "         'tool registry']}]},\n",
       "     'services registry',\n",
       "     'standards registry',\n",
       "     'database registry',\n",
       "     {'data catalogue': [{'synonyms': ['metadata repository',\n",
       "         'data registry',\n",
       "         'data catalog']}]}]},\n",
       "   {'database (tag)': [{'synonyms': ['database portal',\n",
       "       'data repository',\n",
       "       'data archive',\n",
       "       'data archives']}]},\n",
       "   {'knowledgebase (tag)': [{'synonyms': ['resource collection']}]}]}]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(TAG_MAPPING, \"r\") as file:\n",
    "    tag_mapping = yaml.safe_load(file)\n",
    "\n",
    "tag_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43, 356)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def unnest(nested, pre=None):\n",
    "    pre = pre[:] if pre else []\n",
    "    if isinstance(nested, dict):\n",
    "        for key, value in nested.items():\n",
    "            if isinstance(value, dict):\n",
    "                for d in unnest(value, pre + [key]):\n",
    "                    yield d\n",
    "            elif isinstance(value, list):\n",
    "                for v in value:\n",
    "                    for d in unnest(v, pre + [key]):\n",
    "                        yield d\n",
    "            else:\n",
    "                yield pre + [key, value]\n",
    "    elif isinstance(nested, list):\n",
    "        for value in nested:\n",
    "            if isinstance(value, dict):\n",
    "                for d in unnest(value, pre):\n",
    "                    yield d\n",
    "            elif isinstance(value, list):\n",
    "                for v in value:\n",
    "                    for d in unnest(v, pre):\n",
    "                        yield d\n",
    "            else:\n",
    "                yield pre + [key, value]\n",
    "    else:\n",
    "        yield pre + [nested]\n",
    "\n",
    "tags = set()\n",
    "term2parents = ddict(set)\n",
    "term2synonyms = ddict(set)\n",
    "for l in unnest(tag_mapping):\n",
    "    _tags = set([i for i in l if i[-5:] == \"(tag)\"])\n",
    "    if len(_tags) > 1:\n",
    "        print(_tags)\n",
    "    tags |= _tags\n",
    "    if \"synonyms\" in l:\n",
    "        term2synonyms[l[-3]].add(l[-1])\n",
    "    for idx, i in enumerate(l):\n",
    "        if i == \"synonyms\":\n",
    "            break\n",
    "        term2parents[i] |= set(l[:idx])\n",
    "\n",
    "synonym2terms = ddict(set)\n",
    "for k, v in term2synonyms.items():\n",
    "    for i in v:\n",
    "        synonym2terms[i].add(k)\n",
    "\n",
    "term2keywords = ddict(set)\n",
    "for k, v in term2parents.items():\n",
    "    term2keywords[k] |= v.union(*[term2synonyms[i] for i in v])\n",
    "\n",
    "len(tags), len(term2keywords) + len([i for j in term2synonyms.values() for i in j])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "be099a71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "documentation !\n",
      "nucleic acid quantitation !\n",
      "nucleic acid quantitation !\n",
      "methyl-seq kit !\n",
      "EMQN !\n",
      "certification !\n",
      "certification !\n",
      "toolbox !\n",
      "certification !\n",
      "workflow !\n",
      "workflow !\n",
      "SRiC !\n",
      "toolkit !\n",
      "BWAmeth !\n",
      "clinical validation !\n",
      "workflow !\n",
      "study design !\n",
      "study design !\n",
      "study design !\n",
      "samples !\n",
      "toolbox !\n",
      "visualisation tool !\n",
      "study management !\n",
      "study management !\n",
      "study management !\n",
      "bioinformatics !\n"
     ]
    }
   ],
   "source": [
    "# iterate through resources\n",
    "for root, dirs, files in os.walk(RESOURCES_DIR):\n",
    "    for name in files:\n",
    "        if not name.endswith(\".yaml\"):\n",
    "            continue\n",
    "        #print(name)\n",
    "\n",
    "        # load test resource and validate\n",
    "        resource = motbxschema.MotbxResource(os.path.join(root, name))\n",
    "        #resource.validate(schema) don't validate in first version\n",
    "        resource_tags = set()\n",
    "        resource_keywords = set()\n",
    "        for i in resource.resource[\"resourceTags\"]:\n",
    "            if f'{i} (tag)' in tags:\n",
    "                # add old tag to new tags\n",
    "                resource_tags.add(i)\n",
    "                # add synonyms defined for this tag as keywords\n",
    "                resource_keywords |= set(term2keywords[f'{i} (tag)'])\n",
    "            elif i in term2parents:\n",
    "                resource_keywords.add(i)\n",
    "                resource_keywords |= set(term2keywords[i])\n",
    "            else:\n",
    "                #print(i)\n",
    "                if i in synonym2terms.keys():\n",
    "                    #print(synonym2terms[i])\n",
    "                    resource_keywords.add(i)\n",
    "                    for t in synonym2terms[i]:\n",
    "                        if t in tags:\n",
    "                            resource_tags.add(t[:-6])\n",
    "                            resource_keywords |= set(term2synonyms[t])\n",
    "                        else:\n",
    "                            resource_keywords.add(t)\n",
    "                else:\n",
    "                    resource_keywords.add(i)\n",
    "        #print(resource_tags)\n",
    "        #pp.pprint(sorted(resource_keywords))\n",
    "        #break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11 (motbxtools)",
   "language": "python",
   "name": "motbxtools"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
