{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "831ae188-c669-4885-88ce-6f4ba7718be8",
   "metadata": {},
   "source": [
    "# Summarise MOTBX resources\n",
    "\n",
    "Convert MOTBX resources (YAML files) to CSV file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "94f35a6e-6ad6-4801-b66f-496efc669f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import os\n",
    "import csv\n",
    "from motbxtools import motbxschema\n",
    "import json\n",
    "import yaml\n",
    "import pprint\n",
    "pp = pprint.PrettyPrinter(indent = 2)\n",
    "\n",
    "CWD = Path.cwd()\n",
    "if CWD.name != \"notebooks\":\n",
    "    print(\"Make sure to run this notebook from the 'notebooks' directory.\")\n",
    "\n",
    "MOTBX_DIR = CWD.parent # home directory of this repository\n",
    "\n",
    "with open(MOTBX_DIR.joinpath(\"resources/MOTBX_version.yaml\"), \"r\") as f:\n",
    "    MOTBX_VERSION = yaml.safe_load(f)\n",
    "\n",
    "# path to directory where resources YAML file are saved\n",
    "RESOURCES_DIR = MOTBX_DIR.joinpath(\"resources/curated\")\n",
    "SUMMARY_DIR = MOTBX_DIR.joinpath(\"resources/summary\")\n",
    "if not os.path.exists(SUMMARY_DIR):\n",
    "    os.mkdir(SUMMARY_DIR)\n",
    "SUMMARY_CSV = SUMMARY_DIR.joinpath(f\"MOTBX_{MOTBX_VERSION['latest']}.csv\")\n",
    "\n",
    "# path to JSON SCHEMA file defining structure of MOTBX resources\n",
    "SCHEMA_JSON = MOTBX_DIR.joinpath(\"schema/motbxschema.json\")\n",
    "\n",
    "TEST_RESOURCES_DIR = MOTBX_DIR.joinpath(\"tests/resources_pass\")\n",
    "TEST_SUMMARY_CSV = MOTBX_DIR.joinpath(\"tests/resources.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a8e82c5-2fc0-4e13-bf4e-dd7d5e58181b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"$schema\": \"https://json-schema.org/draft/2020-12/schema\",\n",
      "  \"title\": \"MOTBX resource\",\n",
      "  \"description\": \"Schema for resources of the EATRIS Multi-omics Toolbox (MOTBX)\",\n",
      "  \"type\": \"object\",\n",
      "  \"properties\": {\n",
      "    \"resourceID\": {\n",
      "      \"type\": \"string\"\n",
      "    },\n",
      "    \"resourceCategory\": {\n",
      "      \"type\": \"string\",\n",
      "      \"enum\": [\n",
      "        \"Epigenomics\",\n",
      "        \"External Quality Assessment\",\n",
      "        \"Genomics\",\n",
      "        \"Internal Quality Control\",\n",
      "        \"Metabolomics\",\n",
      "        \"Omics data management and analysis\",\n",
      "        \"Proteomics\",\n",
      "        \"Transcriptomics\"\n",
      "      ]\n",
      "    },\n",
      "    \"resourceSubcategory\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 4,\n",
      "      \"maxLength\": 30\n",
      "    },\n",
      "    \"resourceTitle\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 15,\n",
      "      \"maxLength\": 160\n",
      "    },\n",
      "    \"resourceDescription\": {\n",
      "      \"type\": \"string\",\n",
      "      \"minLength\": 50,\n",
      "      \"maxLength\": 2500\n",
      "    },\n",
      "    \"resourceUrl\": {\n",
      "      \"type\": \"string\",\n",
      "      \"format\": \"uri\",\n",
      "      \"pattern\": \"^https://|^http://|^ftp://|.pdf$\"\n",
      "    },\n",
      "    \"resourceTags\": {\n",
      "      \"type\": \"array\",\n",
      "      \"items\": {\n",
      "        \"type\": \"string\"\n",
      "      },\n",
      "      \"minItems\": 1\n",
      "    }\n",
      "  },\n",
      "  \"required\": [\n",
      "    \"resourceID\",\n",
      "    \"resourceCategory\",\n",
      "    \"resourceSubcategory\",\n",
      "    \"resourceTitle\",\n",
      "    \"resourceDescription\",\n",
      "    \"resourceUrl\",\n",
      "    \"resourceTags\"\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# load JSON schema from file\n",
    "schema = motbxschema.MotbxSchema(SCHEMA_JSON)\n",
    "\n",
    "# print schema formatted as JSON\n",
    "print(json.dumps(schema.schema, indent = 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "93013f06-ff43-4643-b860-585e0d0026fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['resourceID', 'resourceCategory', 'resourceSubcategory', 'resourceTitle', 'resourceDescription', 'resourceUrl', 'resourceTags']\n"
     ]
    }
   ],
   "source": [
    "# define CSV column names\n",
    "fieldnames = list(schema.schema[\"properties\"].keys())\n",
    "print(fieldnames)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b50c1c5",
   "metadata": {},
   "source": [
    "Create YAML file summary for test resources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c6a888de-4495-4a5a-b191-c5c7fa3f8685",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test1.yaml\n",
      "{ 'resourceCategory': 'Internal Quality Control',\n",
      "  'resourceDescription': 'ISO Guide 80:2014 guidance for the in-house '\n",
      "                         'preparation of quality control materials (QCMs). ISO '\n",
      "                         'Guide 80 outlines the characteristics and '\n",
      "                         'preparation processes of reference materials for '\n",
      "                         'quality control. It applies to stable materials used '\n",
      "                         'locally and those transported without significant '\n",
      "                         'property changes. Laboratory staff preparing '\n",
      "                         'in-house quality control materials should follow ISO '\n",
      "                         'Guides 34 and 35 for transportation-based supply '\n",
      "                         'chains. The preparation of quality control materials '\n",
      "                         'requires assessments for homogeneity, stability, and '\n",
      "                         'limited characterization. It aims to demonstrate '\n",
      "                         'statistical control in a measurement system but does '\n",
      "                         'not provide usage guidance. The guide offers general '\n",
      "                         'information on preparation and includes case studies '\n",
      "                         'for different sectors. Users should have material '\n",
      "                         'knowledge and be aware of matrix effects and '\n",
      "                         'contamination risks.',\n",
      "  'resourceID': '1',\n",
      "  'resourceSubcategory': 'Guidelines and best practices',\n",
      "  'resourceTags': 'ISO standard, guidelines, quality control material, '\n",
      "                  'in-house, genomics',\n",
      "  'resourceTitle': 'ISO Guide 80:2014: Guidance for in-house preparation of '\n",
      "                   'quality control materials',\n",
      "  'resourceUrl': 'https://www.iso.org/standard/44313.html'}\n",
      "test2.yaml\n",
      "{ 'resourceCategory': 'Proteomics',\n",
      "  'resourceDescription': 'A freely-available and open source Windows client '\n",
      "                         'application for building Selected Reaction '\n",
      "                         'Monitoring (SRM)/Multiple Reaction Monitoring (MRM), '\n",
      "                         'Parallel Reaction Monitoring (PRM), Data Independent '\n",
      "                         'Acquisition (DIA/SWATH) and DDA with MS1 '\n",
      "                         'quantitative methods and analysing the resulting '\n",
      "                         'mass spectrometer data. It aims to employ '\n",
      "                         'cutting-edge technologies for creating and '\n",
      "                         'iteratively refining targeted methods for '\n",
      "                         'large-scale quantitative mass spectrometry studies '\n",
      "                         'in life sciences.',\n",
      "  'resourceID': 'test2',\n",
      "  'resourceSubcategory': 'Raw data processing',\n",
      "  'resourceTags': 'software, data processing, toolbox, mass spectrometry, '\n",
      "                  'proteomics',\n",
      "  'resourceTitle': 'Raw data processing',\n",
      "  'resourceUrl': 'https://skyline.ms/project/home/software/skyline/begin.view'}\n"
     ]
    }
   ],
   "source": [
    "# open CSV file\n",
    "with open(TEST_SUMMARY_CSV, \"w\", newline = \"\", encoding = \"utf-8\") as csvfile:\n",
    "    writer = csv.DictWriter(csvfile, fieldnames = fieldnames)\n",
    "    writer.writeheader()\n",
    "\n",
    "    # iterate through resources\n",
    "    for root, dirs, files in os.walk(TEST_RESOURCES_DIR):\n",
    "        for name in files:\n",
    "            if not name.endswith(\".yaml\"):\n",
    "                continue\n",
    "            print(name)\n",
    "\n",
    "            # load test resource and validate\n",
    "            resource = motbxschema.MotbxResource(os.path.join(root, name))\n",
    "            resource.validate(schema)\n",
    "\n",
    "            # write to CSV file\n",
    "            row = resource.flatten(fieldnames)\n",
    "            pp.pprint(row)\n",
    "            writer.writerow(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0890640e",
   "metadata": {},
   "source": [
    "Create summary for curated resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "be099a71",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID0001.yaml\n",
      "ID0002.yaml\n",
      "ID0003.yaml\n",
      "ID0004.yaml\n",
      "ID0005.yaml\n",
      "ID0006.yaml\n",
      "ID0007.yaml\n",
      "ID0008.yaml\n",
      "ID0009.yaml\n",
      "ID0010.yaml\n",
      "ID0011.yaml\n",
      "ID0012.yaml\n",
      "ID0013.yaml\n",
      "ID0014.yaml\n",
      "ID0015.yaml\n",
      "ID0016.yaml\n",
      "ID0017.yaml\n",
      "ID0018.yaml\n",
      "ID0019.yaml\n",
      "ID0020.yaml\n",
      "ID0021.yaml\n",
      "ID0022.yaml\n",
      "ID0023.yaml\n",
      "ID0024.yaml\n",
      "ID0025.yaml\n",
      "ID0026.yaml\n",
      "ID0027.yaml\n",
      "ID0028.yaml\n",
      "ID0029.yaml\n",
      "ID0030.yaml\n",
      "ID0031.yaml\n",
      "ID0032.yaml\n",
      "ID0033.yaml\n",
      "ID0034.yaml\n",
      "ID0035.yaml\n",
      "ID0036.yaml\n",
      "ID0037.yaml\n",
      "ID0038.yaml\n",
      "ID0039.yaml\n",
      "ID0040.yaml\n",
      "ID0041.yaml\n",
      "ID0042.yaml\n",
      "ID0043.yaml\n",
      "ID0044.yaml\n",
      "ID0045.yaml\n",
      "ID0046.yaml\n",
      "ID0047.yaml\n",
      "ID0048.yaml\n",
      "ID0049.yaml\n",
      "ID0050.yaml\n",
      "ID0051.yaml\n",
      "ID0052.yaml\n",
      "ID0053.yaml\n",
      "ID0054.yaml\n",
      "ID0055.yaml\n",
      "ID0056.yaml\n",
      "ID0057.yaml\n",
      "ID0058.yaml\n",
      "ID0059.yaml\n",
      "ID0060.yaml\n",
      "ID0061.yaml\n",
      "ID0062.yaml\n",
      "ID0063.yaml\n",
      "ID0064.yaml\n",
      "ID0065.yaml\n",
      "ID0066.yaml\n",
      "ID0067.yaml\n",
      "ID0068.yaml\n",
      "ID0069.yaml\n",
      "ID0070.yaml\n",
      "ID0071.yaml\n",
      "ID0072.yaml\n",
      "ID0073.yaml\n",
      "ID0074.yaml\n",
      "ID0075.yaml\n",
      "ID0076.yaml\n",
      "ID0077.yaml\n",
      "ID0078.yaml\n",
      "ID0079.yaml\n",
      "ID0080.yaml\n",
      "ID0081.yaml\n"
     ]
    }
   ],
   "source": [
    "# open CSV file\n",
    "with open(SUMMARY_CSV, \"w\", newline = \"\", encoding = \"utf-8\") as csvfile:\n",
    "    writer = csv.DictWriter(csvfile, fieldnames = fieldnames)\n",
    "    writer.writeheader()\n",
    "\n",
    "    # iterate through resources\n",
    "    for root, dirs, files in os.walk(RESOURCES_DIR):\n",
    "        for name in files:\n",
    "            if not name.endswith(\".yaml\"):\n",
    "                continue\n",
    "            print(name)\n",
    "\n",
    "            # load test resource and validate\n",
    "            resource = motbxschema.MotbxResource(os.path.join(root, name))\n",
    "            #resource.validate(schema) don't validate in first version\n",
    "\n",
    "            # write to CSV file\n",
    "            row = resource.flatten(fieldnames)\n",
    "            writer.writerow(row) # TODO: fix 403 error"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11 (motbxtools)",
   "language": "python",
   "name": "motbxtools"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
